File: netapp-manageability-sdk-9.3/src/sample/Data_ONTAP/Perl/snapman.pl

#============================================================#
#							     *
# $Id$
#                                                            #
# snapman.pl                                                 #
#                                                            #
# Snapshot management using ONTAPI interfaces in Perl        #
#                                                            #
# Copyright 2002-2003 Network Appliance, Inc. All rights     #
# reserved. Specifications subject to change without notice. #
#                                                            #
# This SDK sample code is provided AS IS, with no support or #
# warranties of any kind, including but not limited to       #
# warranties of merchantability or fitness of any kind,      #
# expressed or implied.  This code is subject to the license #
# agreement that accompanies the SDK.                        #
#                                                            #
# tab size = 8                                               #
#                                                            #
#============================================================#

my $VERSION = '1.0';	# Controls the SDK release.

use strict;

use lib '../../../../lib/perl/NetApp';
use NaServer;
use Pod::Usage;

die pod2usage( 
	verbose => 1
) unless $ARGV[4];

our $opt = $ARGV[0];
our $filer = $ARGV[1];
our $user = $ARGV[2];
our $password = $ARGV[3];
our $vol = $ARGV[4];

our $s = NaServer->new($filer, 1, 1);
$s->set_admin_user($user, $password);

#
# snapshot-get-schedule
#
if ($opt eq "-g") {
	my $output = $s->invoke("snapshot-get-schedule", 
		"volume", $vol);
	if ($output->results_errno != 0) {
		my $r = $output->results_reason();
		print "snapshot-get-schedule failed: $r\n";
	}

	my $minutes = $output->child_get_int("minutes", 0);
	my $hours =   $output->child_get_int("hours", 0);
	my $days =    $output->child_get_int("days", 0);
	my $weeks =   $output->child_get_int("weeks", 0);

	my $whichhours = $output->child_get_string("which-hours");
	my $whichminutes = $output->child_get_string("which-minutes");

	printf("\n");
	printf("Snapshot schedule for volume %s on filer %s:\n", $vol, $filer);
	printf("\n");
	if ($minutes > 0) {
		printf("Snapshots are taken on minutes [%s] of each hour (%d kept)\n",
			$whichminutes, $minutes);
	}
	if ($hours > 0) {
		printf("Snapshots are taken on hours [%s] of each day (%d kept)\n", 
			$whichhours, $hours);
	}
	if ($days > 0) {
		printf("%d nightly snapshots are kept\n", $days);
	}
	if ($weeks) {
		printf("%d weekly snapshots are kept\n", $weeks);
	}

	if ($minutes == 0 && $hours == 0 && $days == 0 && $weeks == 0) {
		printf("No snapshot schedule\n");
	}
	printf("\n");
}

#
# snapshot-list-info
#
elsif ($opt eq "-l") {
	my $output = $s->invoke("snapshot-list-info", "volume", $vol);

	if ($output->results_errno != 0) {
		my $r = $output->results_reason();
		print "snapshot-list-info failed: $r\n";
	}

	# 
	# get snapshot list
	#
	my $snapshotlist = $output->child_get("snapshots");
	if (!defined($snapshotlist) || ($snapshotlist eq "")) {
		# no snapshots to report
		printf("No snapshots on volume %s \n\n", $vol);
		exit(0);
	}

	#
	# print header
	#
	printf("Snapshots on volume %s: \n\n", $vol);         
	printf("NAME                    DATE                    BUSY   NBLOCKS CUMNBLOCKS  DEPENDENCY\n");
	printf("-------------------------------------------------------------------------------------\n");

	# 
	# iterate through snapshot list
	#
	my @snapshots = $snapshotlist->children_get();
	foreach my $ss (@snapshots) {

		my $accesstime = $ss->child_get_int("access-time", 0);
		my $total =		 $ss->child_get_int("total", 0);
		my $cumtotal =   $ss->child_get_int("cumulative-total", 0);

		my $busy = ($ss->child_get_string("busy") eq "true");
		
		my $dependency = $ss->child_get_string("dependency");
		my $name =		 $ss->child_get_string("name");

		my $date = localtime($accesstime);

		#
		# print a line
		#
		printf("%-23s %-24s  %d %10d %10d  %s\n",
				$name,
				$date,
				$busy,
				$total,
				$cumtotal,
				$dependency); 
	}
}

#
# snapshot-create
#
elsif ($opt eq "-c") {
	
	die pod2usage( 
		verbose => 1
	) unless $ARGV[5];

	my $name = $ARGV[5];

	my $output = $s->invoke("snapshot-create",
				"volume", $vol,
				"snapshot", $name);

	if ($output->results_errno != 0) {
		my $r = $output->results_reason();
		print "snapshot-create failed: $r\n";
	}
}

#
# snapshot-rename
#
elsif ($opt eq "-r") {

	die pod2usage( 
		verbose => 1
	) unless $ARGV[6];

	my $currname = $ARGV[5];
	my $newname = $ARGV[6];

	my $output = $s->invoke("snapshot-rename",
				"volume", $vol,
				"current-name", $currname,
				"new-name", $newname);

	if ($output->results_errno != 0) {
		my $r = $output->results_reason();
		print "snapshot-rename failed: $r\n";
	}
}

#
# snapshot-delete
#
elsif ($opt eq "-d") {

	die pod2usage( 
		verbose => 1
	) unless $ARGV[5];

	my $name = $ARGV[5];

	my $output = $s->invoke("snapshot-delete",
				"volume", $vol,
				"snapshot", $name);

	if ($output->results_errno != 0) {
		my $r = $output->results_reason();
		print "snapshot-rename failed: $r\n";
	}
}

else {
	
	die pod2usage(verbose => 1);
}


#=========================== POD ============================#

=head1 NAME

  snapman.pl - Snapshot management using ONTAPI

=head1 SYNOPSIS

  snapman.pl <filer name> <user> <password>
  snapman.pl -g <filer> <user> <pw> <vol> 
			 -l <filer> <user> <pw> <vol> 
			 -c <filer> <user> <pw> <vol> <snapshotname> 
			 -r <filer> <user> <pw> <vol> <oldsnapshotname> <newname> 
			 -d <filer> <user> <pw> <vol> <snapshotname>
			 
  E.g. snapman.pl -l filer1 root 6a55w0r9 vol0

  Use -g to get the snapshot schedule
	  -l to list snapshot info 
	  -c to create a snapshot 
	  -r to rename one 
	  -d to delete one 


=head1 ARGUMENTS

  <option>
  -g, -l, -c, -r, or -d
  
  <filer Name>
  The name of the filer to test.

  <user>
  username.

  <password>
  password.

  <vol>
  Name of volume for this operation

  <snapshotname>
  Name of snapshot to create or delete

  <oldsnapshotname>
  Name of snapshot to rename

  <newsnapshotname>
  Name to rename snapshot to


=head1 SEE ALSO

  NaElement.pm, NaServer.pm

=head1 COPYRIGHT

  Copyright 2002-2003 Network Appliance, Inc. All rights 
  reserved. Specifications subject to change without notice.

  This SDK sample code is provided AS IS, with no support or 
  warranties of any kind, including but not limited to 
  warranties of merchantability or fitness of any kind, 
  expressed or implied.  This code is subject to the license 
  agreement that accompanies the SDK.

=cut

File: netapp-manageability-sdk-9.3/src/sample/Data_ONTAP/DotNet/VB.NET/snapman/snapman.vb
Imports System
Imports System.Collections.Generic
Imports NetApp.Manage

Module SnapMan
    Public Sub Usage()
        Console.WriteLine(Environment.NewLine & "Usage: snapman <-list|-create|-rename|-delete> <filer> <user> <passwd> <vol> ")
        Console.WriteLine(Environment.NewLine & " -list <filer> <user> <passwd> <vol> ")
        Console.WriteLine(" -create <filer> <user> <passwd> <vol> <snapshotname> ")
        Console.WriteLine(" -rename <filer> <user> <passwd> <vol> <oldsnapshotname> <newname>")
        Console.WriteLine(" -delete <filer> <user> <passwd> <vol> <snapshotname>")
        Console.WriteLine(" -schedule <filer> <user> <passwd> <vol>")
        Console.WriteLine(Environment.NewLine & "E.g. snapman -list filer1 root 6a55w0r9 vol0")
        Environment.Exit(-1)
    End Sub

    Public Sub CreateSnapshot(ByVal args() As String)
        Dim serverName As String = args(1)
        Dim user As String = args(2)
        Dim passwd As String = args(3)
        Dim volume As String = args(4)
        Dim snapshot As String = ""
        Dim server As NaServer
        Dim input As NaElement
        Dim output As NaElement

        Try

            server = New NaServer(serverName, 1, 0)
            server.SetAdminUser(user, passwd)

            input = New NaElement("snapshot-create")
            If args.Length = 6 Then
                snapshot = args(5)
                input.AddNewChild("volume", volume)
                input.AddNewChild("snapshot", snapshot)
            Else
                Console.Error.WriteLine("Invalid number of arguments")
                Usage()
                Environment.Exit(-1)
            End If
            output = server.InvokeElem(input)
            Console.WriteLine("Snapshot " + snapshot + " created for volume " + volume + " on filer " + serverName)
        Catch ex As Exception
            Console.Error.WriteLine("ERROR: " + ex.Message)
        End Try
    End Sub
    Public Sub DeleteSnapshot(ByVal args() As String)

        Dim serverName As String = args(1)
        Dim user As String = args(2)
        Dim passwd As String = args(3)
        Dim volume As String = args(4)
        Dim snapshot As String = ""
        Dim server As NaServer
        Dim input As NaElement
        Dim output As NaElement

        Try
            server = New NaServer(serverName, 1, 0)
            server.SetAdminUser(user, passwd)
            input = New NaElement("snapshot-delete")

            If args.Length = 6 Then
                snapshot = args(5)
                input.AddNewChild("volume", volume)
                input.AddNewChild("snapshot", snapshot)
            Else
                Console.Error.WriteLine("Invalid number of arguments")
                Usage()
                Environment.Exit(-1)
            End If
            output = server.InvokeElem(input)
            Console.WriteLine("Snapshot " + snapshot + " deleted from volume " + volume + " on filer " + serverName)

        Catch ex As Exception
            Console.Error.WriteLine("ERROR: " + ex.Message)
        End Try
    End Sub
    Public Sub RenameSnapshot(ByVal args() As String)

        Dim serverName As String = args(1)
        Dim user As String = args(2)
        Dim passwd As String = args(3)
        Dim volume As String = args(4)
        Dim oldSnapshot As String = ""
        Dim newSnapshot As String = ""
        Dim server As NaServer
        Dim input As NaElement
        Dim output As NaElement

        Try
            server = New NaServer(serverName, 1, 0)
            server.SetAdminUser(user, passwd)
            input = New NaElement("snapshot-rename")

            If args.Length = 7 Then
                oldSnapshot = args(5)
                newSnapshot = args(6)
                input = New NaElement("snapshot-rename")
                input.AddNewChild("volume", volume)
                input.AddNewChild("current-name", oldSnapshot)
                input.AddNewChild("new-name", newSnapshot)
            Else
                Console.Error.WriteLine("Invalid number of arguments")
                Usage()
                Environment.Exit(-1)
            End If

            output = server.InvokeElem(input)
            Console.WriteLine("Snapshot " + oldSnapshot + " renamed to " + newSnapshot + " for volume " + volume + " on filer " + serverName)
        Catch ex As Exception
            Console.Error.WriteLine("ERROR: " + ex.Message)
        End Try
    End Sub
    Public Sub ListInfo(ByVal args() As String)

        Dim serverName As String = args(1)
        Dim user As String = args(2)
        Dim passwd As String = args(3)
        Dim volume As String = args(4)
        Dim server As NaServer
        Dim input As NaElement
        Dim output As NaElement

        Try

            server = New NaServer(serverName, 1, 0)
            server.SetAdminUser(user, passwd)
            input = New NaElement("snapshot-list-info")
            input.AddNewChild("volume", volume)
            output = server.InvokeElem(input)

            Dim snapshots As IList = output.GetChildByName("snapshots").GetChildren()
            Dim snapiter As IEnumerator = snapshots.GetEnumerator()
            While snapiter.MoveNext()

                Dim snapshot As NaElement = snapiter.Current
                Console.WriteLine("SNAPSHOT:")
                Dim accessTime As Integer = snapshot.GetChildIntValue("access-time", 0)
                Dim dateTime As DateTime = New DateTime(1970, 1, 1, 0, 0, 0).AddSeconds(accessTime)
                Console.WriteLine(" NAME=" + snapshot.GetChildContent("name"))
                Console.WriteLine(" ACCESS TIME (GMT)=" + dateTime)
                Console.WriteLine(" BUSY= " + snapshot.GetChildContent("busy"))
                Console.WriteLine(" TOTAL (of 1024B)= " + snapshot.GetChildContent("total"))
                Console.WriteLine(" CUMULATIVE TOTAL (of 1024B)= " + snapshot.GetChildContent("cumulative-total"))
                Console.WriteLine(" DEPENDENCY= " + snapshot.GetChildContent("dependency"))
                Console.WriteLine()
            End While
        Catch ex As Exception
            Console.Error.WriteLine("ERROR: " + ex.Message)
        End Try

    End Sub
    Public Sub GetSchedule(ByVal args() As String)

        Dim serverName As String = args(1)
        Dim user As String = args(2)
        Dim passwd As String = args(3)
        Dim volume As String = args(4)
        Dim server As NaServer
        Dim input As NaElement
        Dim output As NaElement

        Try

            server = New NaServer(serverName, 1, 0)
            server.SetAdminUser(user, passwd)

            input = New NaElement("snapshot-get-schedule")
            If args.Length = 5 Then

                input.AddNewChild("volume", volume)

            Else

                Console.Error.WriteLine("Invalid number of arguments")
                Usage()
                Environment.Exit(-1)
            End If

            output = server.InvokeElem(input)
            Console.WriteLine("Snapshot schedule for volume " + volume + " on filer " + serverName)
            Console.WriteLine("---------------------------------------------------------------")
            Console.WriteLine("Snapshots are taken on minutes [" + output.GetChildContent("which-minutes") + "] of each hour (" + output.GetChildContent("minutes") + " kept)")
            Console.WriteLine("Snapshots are taken on hours [" + output.GetChildContent("which-hours") + "] of each day (" + output.GetChildContent("hours") + " kept)")
            Console.WriteLine(output.GetChildContent("days") + " nightly snapshots are kept")
            Console.WriteLine(output.GetChildContent("weeks") + " weekly snapshots are kept")
            Console.WriteLine("")

        Catch ex As Exception
            Console.Error.WriteLine("ERROR: " + ex.Message)
        End Try
    End Sub



    Sub Main(ByVal args As String())

        If (args.Length < 5) Then
            Usage()
            Exit Sub
        End If

        Select Case args(0)
            Case "-create"
                CreateSnapshot(args)
            Case "-rename"
                RenameSnapshot(args)
            Case "-delete"
                DeleteSnapshot(args)
            Case "-list"
                ListInfo(args)
            Case "-schedule"
                GetSchedule(args)
            Case Else
                Usage()
        End Select
    End Sub
End Module

